go/vt/wrangler/reparent.go

```diff
@@ -599,11 +607,34 @@ func (wr *Wrangler) plannedReparentShardLocked(ctx context.Context, ev *events.R
 		wr.logger.Infof("promote replica %v", masterElectTabletAliasStr)
 		event.DispatchUpdate(ev, "promoting replica")
 
+		waitCtx, waitCancel := context.WithTimeout(ctx, waitReplicasTimeout)
+		defer waitCancel()
+
+		err = wr.tmc.WaitForPosition(waitCtx, masterElectTabletInfo.Tablet, rp)
+
+		if err != nil || ctx.Err() != nil && ctx.Err() == context.DeadlineExceeded {
+			// If the new master fails to catch up within the timeout,
+			// we try to roll back to the original master before aborting.
+			// It is possible that we have used up the original context, or that
+			// not enough time is left on it before it times out.
+			// But at this point we really need to be able to Undo so as not to
+			// leave the cluster in a bad state.
+			// So we create a fresh context based on context.Background().
+			undoCtx, undoCancel := context.WithTimeout(context.Background(), *topo.RemoteOperationTimeout)
+			defer undoCancel()
+			if err1 := wr.tmc.UndoDemoteMaster(undoCtx, oldMasterTabletInfo.Tablet); err1 != nil {
+				log.Warningf("Encountered error %v while trying to undo DemoteMaster", err1)
+			}
+			if err != nil {
+				return fmt.Errorf("master-elect tablet %v failed to catch up with replication: %v", masterElectTabletAliasStr, err)
+			}
+			return vterrors.New(vtrpcpb.Code_DEADLINE_EXCEEDED, "PlannedReparent timed out, please try again.")
+		}
+
 		promoteCtx, promoteCancel := context.WithTimeout(ctx, waitReplicasTimeout)
 		defer promoteCancel()
-
-		rp, err = wr.tmc.PromoteSlaveWhenCaughtUp(promoteCtx, masterElectTabletInfo.Tablet, rp)
-		if err != nil || (ctx.Err() != nil && ctx.Err() == context.DeadlineExceeded) {
+		rp, err = wr.tmc.PromoteReplica(promoteCtx, masterElectTabletInfo.Tablet)
+		if err != nil {
```

Comment:
Isn't this the one where we can't be sure it's ok to undo, so we decided to require that they run PRS again to either retry or roll back?

Reply:
I was hoping that an error from `PromoteReplica` meant that the replica is still read-only. But looking at the code, that is not the last thing we do, so we can't be sure.
Fixed to not call Undo here any more.